src/libmeasurement_kit/ooni/bouncer_impl.hpp

```diff
@@ -0,0 +1,124 @@
+                    Callback<Error, Var<Reply>> cb, Settings settings,
+                    Var<Reactor> reactor, Var<Logger> logger) {
+
+    json request;
+    json net_tests;
+    net_tests["input-hashes"] = {};
+    net_tests["name"] = test_name;
+    net_tests["test-helpers"] = helpers;
+    net_tests["version"] = test_version;
+    request["net-tests"][0] = net_tests;
+
+    settings["http/url"] = base_bouncer_url;
+    settings["http/method"] = "POST";
+
+    http_request(
+        settings, {}, request.dump(),
+        [=](Error e, Var<http::Response> resp) {
+            if (e) {
+                reactor->call_soon([=]() { cb(e, nullptr); });
+            }
+            Var<Reply> reply(new Reply);
+            try {
+                json response = json::parse(resp->body);
+                if (response.find("error") != response.end()) {
+                    if (response["error"] == "collector-not-found") {
+                        logger->warn(
+                            "collector not found for the requested test");
+                        reactor->call_soon([=]() {
+                            cb(BouncerCollectorNotFoundError(), nullptr);
+                        });
+                        return;
+                    } else if (response["error"] == "invalid-request") {
+                        logger->warn("invalid request sent to the bouncer");
+                        reactor->call_soon([=]() {
+                            cb(BouncerInvalidRequestError(), nullptr);
+                        });
+                        return;
+                    } else {
+                        // I assume that if we receive a json with the key
+                        // "error"
+                        // then the json is not valid and we should not continue
+                        // to parse it
+                        logger->warn("bouncer generic error");
+                        reactor->call_soon(
+                            [=]() { cb(BouncerGenericError(), nullptr); });
+                        return;
+                    }
+                }
+
+                auto collector_alternate =
+                    response["net-tests"][0]["collector-alternate"];
+                bool collector_found = false;
+                for (auto collector : collector_alternate) {
+                    if (collector["type"] == "https") {
+                        reply->https_collector = collector["address"];
+                        collector_found = true;
+                        break;
+                    }
+                }
+                if (!collector_found) {
+                    logger->warn(
+                        "no https collector found in bouncer response");
+                    reactor->call_soon(
+                        [=]() { cb(MissingHttpsCollectorError(), nullptr); });
+                }
+                auto test_helpers_alternate =
+                    response["net-tests"][0]["test-helpers-alternate"]
```

Comment:
You probably want to also check to see if the array `response["net-tests"]` has a non zero length and fail gracefully in that case.